src/module.c

```diff
@@ -3047,6 +3050,76 @@ int RM_HashGet(RedisModuleKey *key, int flags, ...) {
     return REDISMODULE_OK;
 }
 
+/* --------------------------------------------------------------------------
+ * Key API for the stream type.
+ * -------------------------------------------------------------------------- */
+
+/**
+ * Adds an entry to a stream. Like XADD without trimming.
+ *
+ * - `key`: The key where the stream is (or will be) stored
+ * - `id`: If the `AUTOID` flag is set, this is where the assigned ID is
+ *   returned. Can be NULL if `AUTOID` is set, if you don't care to receive the
+ *   ID. If `AUTOID` is not set, this is the requested ID.
+ * - `flags`: A bit field of
+ *   - `REDISMODULE_STREAM_AUTOID`: Assign a stream ID automatically, like `*`
+ *     in the XADD command.
+ *   - `REDISMODULE_STREAM_NOMKSTREAM`: Don't create the stream if it doesn't
+ *     exist.
+ * - `argv`: A pointer to an array of size `numfields * 2` containing the
+ *   fields and values.
+ * - `numfields`: The number of field-value pairs in `argv`.
+ *
+ * Returns `REDISMODULE_OK` if an entry has been added. Returns
+ * `REDISMODULE_ERR` if the key is not opened for writing, if the key has a type
+ * other than stream, if the key doesn't exist in case the flag `NOMKSTREAM` is
+ * set or when attempting to create an entry with an id which is not allowed.
+ */
+int RM_StreamAdd(RedisModuleKey *key, RedisModuleStreamID *id, int flags,
+                 RedisModuleString **argv, int64_t numfields) {
+    /* Validate args */
+    if (!(key->mode & REDISMODULE_WRITE))
+        return REDISMODULE_ERR;
+    if (key->value && key->value->type != OBJ_STREAM)
+        return REDISMODULE_ERR;
+    if (key->value == NULL && (flags & REDISMODULE_STREAM_NOMKSTREAM))
+        return REDISMODULE_ERR;
+    if (!(flags & REDISMODULE_STREAM_AUTOID) &&
+        (id == NULL || (id->ms == 0 && id->seq == 0)))
+        return REDISMODULE_ERR;
+
+    /* Create key if necessery */
+    if (key->value == NULL)
+        moduleCreateEmptyKey(key, REDISMODULE_KEYTYPE_STREAM);
+
+    stream *s = key->value->ptr;
+
+    /* We don't assume streamID and RedisModuleStreamID are identical.
+       Should we? */
+    streamID added_id;
+    streamID use_id;
+    streamID *use_id_ptr = NULL;
+    if (!(flags & REDISMODULE_STREAM_AUTOID)) {
+        use_id.ms = id->ms;
+        use_id.seq = id->seq;
+        use_id_ptr = &use_id;
+    }
+    if (streamAppendItem(s, argv, numfields, &added_id, use_id_ptr) == C_ERR) {
+        /* Current last ID is greater than or equal to 'use_id' */
+        return REDISMODULE_ERR;
```

Comment:
we might want to give the user some visibility on the error types.
RM_Call does it with `errno`, i'm not sure if we wanna keep that practice.
need to think this through

Reply:
Another idea is define more non-zero return values to use only in new API functions, e.g. `REDISMODULE_ERR_BADARG`, `REDISMODULE_ERR_BADTYPE`, etc. We can't add them to old functions though since users may have assumed that `result != REDISMODULE_ERR` means success.

But, given that it's is possible to prevent all the errors by properly opening the key for writing and checking the type of the key before calling StreamAdd with the right args, I think just returning ERR is OK.

If we return `REDISMODULE_ERR` for now, we can add errno later if you decide upon that?

Reply:
i vote for defining more non-zero errors
AFAIC i don't mind having specific errors for specific APIs as long as it's documented
Otherwise maybe we can find some general errors that are not API-specific

Reply:
Trying to come up with generic `REDISMODULE_ERR_*` usable also for future APIs, here are some far-from-perfect ideas. I think errno is not very well suited in some cases, but I've tried to map them too.

| Scenario | Functions | errno | Return macro idea |
|-|-|-|-|
| invalid flags | Add, IteratorStart, Trim* | EINVAL | `REDISMODULE_ERR_INVAL` |
| key is NULL   | Add, Iterator*, Trim* | EINVAL | `REDISMODULE_ERR_INVAL` |
| argv is NULL  | Add, IteratorNext (when numfields > 0) | EINVAL | `REDISMODULE_ERR_INVAL` |
| id is NULL    | Add (without AUTOID flag) | EINVAL | `REDISMODULE_ERR_INVAL` |
| key not open for writing | Add, Trim* | EBADF | `REDISMODULE_ERR_READONLY` |
| key not existing | Add (with NOMKSTREAM), Iterator*, Trim* | ENOENT | `REDISMODULE_ERR_NOTEXIST` |
| key not a stream (wrong type) | Add, Iterator*, Trim* | ENOTSUP | `REDISMODULE_ERR_TYPE` |
| requested ID is 0-0 | Add | EDOM | `REDISMODULE_ERR_TOOSMALL` |
| requested ID not greater than last stream ID | Add | EDOM | `REDISMODULE_ERR_TOOSMALL` |
| stream exhausted, ID reached the maximum | Add | EFBIG, ENOSPC | `REDISMODULE_ERR_FULL` |
| iterator has no more elements | IteratorNext | ENOMSG | `REDISMODULE_EOF` (not an error) |
| iterator not started | IteratorNext, IteratorStop | EBADF | `REDISMODULE_ERR_NOTSTARTED` |
| string not a stream ID | StringToStreamID | EDOM | `REDISMODULE_ERR` (aligned with existing string functions) |

Most of the errors are programming errors. Maybe it could be possible to enable debug logging instead and get a readable error message printed somewhere, such as by putting a log function pointer in the context or so?

The only errors which aren't possible to detect in other ways are "iterator reaches end of range" (not an error) and "stream ID not available for add" (can be indicated by setting `id` to the next available ID while returning an error). Maybe we can have special return values for only these (and maybe a few more) and generic ERR for the rest?

Reply:
Indeed we can't change existing APIs, since some module can explicitly check for ERR return. 
And introducing new error types just for these APIs is bad for consistency. 
So I tends to lean towards using `errno`.
I'd like to hear what @yossigo thinks.

Reply:
As much as I REALLY don't like the `errno` abuse, I think at this point in the API's life cycle we should probably stick to it rather than create a new error coding system. I think the cognitive load of remembering when to extended `REDISMODULE_ERR_*` and when to use `errno` will inevitably lead to unnecessary bugs.

Reply:
@zuiderkwast i didn't read the entire correspondence - what's the decision here?

Reply:
i think we decided to go with `errno`. Yossi's last post sums it up nicely

Reply:
I'll go with `errno`. I think it's quite fine in all cases except when an iterator comes to its end. EOF is not an error. I've considered ENOENT for "no more entries" but you shouldn't need to check errno in a successful iteration, right?

How about returning OK (0), ERR (1) or EOF (-1) for IteratorNext?

Reply:
i don't like the -1 that much.
considering the user wants to abort the iterations on EOF, i guess it's not too bad to return ERR, with errno set to ENOENT

Reply:
OK @oranagra, I don't like either of the two solutions actually. :-) I'd like to think about it a bit more to see if there are any other options. Any ideas are welcome.

I pushed two commits just now. If we decide to go with ENOENT, I'll revert the last one.

Reply:
@yossigo what do you think? when an iterator is spent, return a REDISMODULE_ERR and errno=ENOENT, or add another REDISMODULE_EOF?

Reply:
@zuiderkwast FYI, in the commit that added the errno, you documented the ENOENT, but didn't actually use it.
so if you revert the last commit, you'll need to fix this.

Reply:
I've been thinking about ERR vs EOF for the IteratorNext() functions a bit more and I think you're right, ERR is cleaner for the user. Motivation: If StreamIteratorStart(key, ...) was successful, IteratorNextID(key, ...) can't fail if it's called with the same key, so in practice, users don't need to check errno for IteratorNextID().

I'll revert to ERR and add a note about this. If @yossigo or anyone else has other ideas, I'll be happy to change it again.

Reply:
@zuiderkwast I agree with you, I think it's not perfect but it's more consistent. In addition to the usage you describe, I think introducing `REDISMODULE_EOF` globally and using it only here would be confusing.